pilot/pkg/serviceregistry/external/servicediscovery.go

```diff
@@ -225,11 +235,82 @@ func getServiceEntryHandler(c *ServiceEntryStore) func(model.Config, model.Confi
 	}
 }
 
+// GetForeignServiceInstanceHandler defines the handler for service instances generated by other registries
+func (d *ServiceEntryStore) GetForeignServiceInstanceHandler() func(*model.ServiceInstance, model.Event) {
+	return func(si *model.ServiceInstance, event model.Event) {
+		key := configKey{
+			kind:      foreignInstanceConfigType,
+			name:      si.Endpoint.Address,
+			namespace: si.Service.Attributes.Namespace,
+		}
+		// Used to indicate if this event was fired for a pod->workloadentry conversion
+		// and that the event can be ignored due to no relevant change in the workloadentry
+		redundantEventForPod := false
+
+		d.storeMutex.Lock()
+		// We will only select entries in the same namespace
+		entries := d.seWithSelectorByNamespace[si.Service.Attributes.Namespace]
+
+		// this is from a pod. Store it in separate map so that
+		// the refreshIndexes function can use these as well as the store ones.
+		switch event {
+		case model.EventDelete:
+			if _, exists := d.foreignRegistryInstancesByIP[si.Endpoint.Address]; !exists {
+				// multiple delete events for the same pod (succeeded/failed/unknown status repeating).
+				redundantEventForPod = true
+			} else {
+				delete(d.foreignRegistryInstancesByIP, si.Endpoint.Address)
+			}
+		default: // add or update
+			if old, exists := d.foreignRegistryInstancesByIP[si.Endpoint.Address]; exists {
+				// If multiple k8s services select the same pod, we may be getting multiple events
+				// ignore them as we only care about the Endpoint itself.
+				if reflect.DeepEqual(old.Endpoint, si.Endpoint) {
+					// ignore the udpate as nothing has changed
+					redundantEventForPod = true
+				}
+			}
+			d.foreignRegistryInstancesByIP[si.Endpoint.Address] = si
+		}
+		d.storeMutex.Unlock()
+
+		// nothing useful to do.
+		if len(entries) == 0 || redundantEventForPod {
+			return
+		}
+
+		log.Debugf("Handle event %s for service instance (from %s) in namespace %s", event,
+			si.Service.Hostname, si.Service.Attributes.Namespace)
+		instances := []*model.ServiceInstance{}
+
+		for _, se := range entries {
+			workloadLabels := labels.Collection{si.Endpoint.Labels}
+			if !workloadLabels.IsSupersetOf(se.entry.WorkloadSelector.Labels) {
+				// Not a match, skip this one
+				continue
+			}
+			instance := convertForeignServiceInstances(si, se.services, se.entry)
+			instances = append(instances, instance...)
+		}
+
+		if event != model.EventDelete {
+			d.updateExistingInstances(key, instances)
+		} else {
+			d.deleteExistingInstances(key, instances)
+		}
+
+		d.edsUpdate(instances)
+		return
+	}
+}
+
 func (d *ServiceEntryStore) Provider() serviceregistry.ProviderID {
 	return serviceregistry.External
 }
 
 func (d *ServiceEntryStore) Cluster() string {
+	// DO NOT ASSIGN CLUSTER ID to non-k8s registries. This will prevent service entries with multiple
```

Comment:
this is just copied from aggregate controller.

Reply:
Interesting. This can be avoided by having different ServiceEntries, one per cluster - since the VMs are treated as extension of a cluster.

@howardjohn  - I don't remember what was the bug that resulted in tracking the cluster. Would it affect VMs ? 

@mandarjog  does it impact telemetry ?

Not related with this PR, but we haven't looked close at this for some time.

Reply:
@douglas-reid i think this should be fine for vm telemetry.

Reply:
Is this impacting the Envoy metadata?  I'm not familiar enough with the use of this code.  In most scenarios, we've avoided giving VMs any info about cluster (as far as telemetry is concerned), so this seems consistent.